src/common/runtime/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+
+#[derive(Debug, Clone, Copy)]
+enum PoolType {
+    Compute,
+    IO,
+}
+
+pub struct Runtime {
+    runtime: tokio::runtime::Runtime,
+    pool_type: PoolType,
+}
+
+impl Runtime {
+    pub(crate) fn new(runtime: tokio::runtime::Runtime, pool_type: PoolType) -> RuntimeRef {
+        Arc::new(Self { runtime, pool_type })
+    }
+
+    async fn execute_task<F>(future: F, pool_type: PoolType) -> DaftResult<F::Output>
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        AssertUnwindSafe(future).catch_unwind().await.map_err(|e| {
+            let s = if let Some(s) = e.downcast_ref::<String>() {
+                s.clone()
+            } else if let Some(s) = e.downcast_ref::<&str>() {
+                (*s).to_string()
+            } else {
+                "unknown internal error".to_string()
+            };
+            DaftError::ComputeError(format!(
+                "Caught panic when spawning blocking task in the {:?} runtime: {})",
+                pool_type, s
+            ))
+        })
+    }
+
+    /// Spawns a task on the runtime and blocks the current thread until the task is completed.
+    /// Similar to tokio's Runtime::block_on but requires static lifetime + Send
+    /// You should use this when you are spawning IO tasks from an Expression Evaluator or in the Executor
+    pub fn block_on<F>(&self, future: F) -> DaftResult<F::Output>
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        let (tx, rx) = oneshot::channel();
+        let pool_type = self.pool_type;
+        let _join_handle = self.spawn(async move {
+            let task_output = Self::execute_task(future, pool_type).await;
+            if tx.send(task_output).is_err() {
+                log::warn!("Spawned task output ignored: receiver dropped");
+            }
+        });
+        rx.recv().expect("Spawned task transmitter dropped")
+    }
+
+    /// Spawn a task on the runtime and await on it.
+    /// You should use this when you are spawning compute or IO tasks from the Executor.
+    pub async fn await_on<F>(&self, future: F) -> DaftResult<F::Output>
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        let (tx, rx) = oneshot::channel();
+        let pool_type = self.pool_type;
+        let _join_handle = self.spawn(async move {
+            let task_output = Self::execute_task(future, pool_type).await;
+            if tx.send(task_output).is_err() {
+                log::warn!("Spawned task output ignored: receiver dropped");
+            }
+        });
+        rx.await.expect("Spawned task transmitter dropped")
+    }
+
+    /// Blocks current thread to compute future. Can not be called in tokio runtime context
+    ///
+    pub fn block_on_current_thread<F: Future>(&self, future: F) -> F::Output {
+        self.runtime.block_on(future)
+    }
+
+    pub fn spawn<F>(&self, future: F) -> JoinHandle<F::Output>
```

Comment:
discussed offline but we should have a way to cancel the Future if this output is dropped.

Reply:
Added todo for this